### PR TITLE
cmake: use the standard CMake flag to drive the shared object build

### PR DESCRIPTION
**Pull request by yegorich**
Remove both POCO_STATIC and LIB_MODE variables and let
the standard CMake flags drive the shared object build.

Build shared library:
cmake -DBUILD_SHARED_LIBS=ON

Build static library:
cmake -DBUILD_SHARED_LIBS=OFF

### CODE REVIEW

**Comment by yegorich** [comment]
I'll send v2 tomorrow fixing tests building.

**Comment by yegorich** [comment]
So Travis failures 3631.1 and 3631.2 have nothing to do with CMake changes but 3631.5. Should I send v3 with changes to .travis.yml like these, that would restore the default build behavior?

index 264252f..8e85697 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -144,7 +144,7 @@ matrix:
         - $CXX --version
         - source ./travis/ignored.sh
         - export POCO_BASE=`pwd`
-        - mkdir cmake-build && cd cmake-build && cmake  -DENABLE_PDF=OFF -DENABLE_TESTS=ON .. && make -s -j2 && ctest -VV -E Data && cd ..
+        - mkdir cmake-build && cd cmake-build && cmake -DBUILD_SHARED_LIBS=ON -DENABLE_PDF=OFF -DENABLE_TESTS=ON .. && make -s -j2 && ctest -VV -E Data && cd ..


     - env:    test="4.0 x86 (make) bundled"
@@ -219,7 +219,7 @@ matrix:
         - source ./travis/ignored.sh
         - export POCO_BASE=`pwd`
         - mkdir cmake-build
-        - cd cmake-build && cmake -DENABLE_PDF=OFF -DENABLE_NETSSL=OFF -DENABLE_CRYPTO=OFF -DENABLE_TESTS=ON .. && make -s -j2 && cd ..
+        - cd cmake-build && cmake -DBUILD_SHARED_LIBS=ON -DENABLE_PDF=OFF -DENABLE_NETSSL=OFF -DENABLE_CRYPTO=OFF -DENABLE_TESTS=ON .. && make -s -j2 && cd ..


     # QA jobs for code analytics and metrics

**Comment by bachp** [comment]
@yegorich From what I understand this will make the default `STATIC` and only build shared if `-DBUILD_SHARED_LIBS=ON` is added?

**Comment by yegorich** [comment]
Yes.

**Comment by yegorich** [comment]
@bachp I've uploaded v3. As stated in the changed commit log, I've introduced `BUILD_SHARED_LIBS` as an option and initialized it as `ON`. So now the default behavior is preserved and CMake tests are successful (at least on Linux).

**Comment by zosrothko** [comment]
@yegorich @bachp The issue below should also be fixed on Travis
```
-- Found MySQL Include directory: /usr/include/mysql  library directory: /usr/lib/x86_64-linux-gnu
-- MySQL Support Enabled
-- Could NOT find PostgreSQL (missing: PostgreSQL_TYPE_INCLUDE_DIR) (found version "10.1")
-- PostgreSQL Support Disabled - no PostgreSQL library
```

**Comment by zosrothko** [comment]
@yegorich @bachp and also those errors on AppVeyor. See [AppVeyor job](https://ci.appveyor.com/project/obiltschnig/poco/build/build%20develop-2685/job/bc6hbvhnx4cwc2u5)

**Comment by yegorich** [comment]
@zosrothko my patch has nothing to do with  PostgreSQL detection. On my system I have no problems finding PostgreSQL 9.6.6 (Debian 9) and 10.1 (Buildroot) with CMake 3.7.2.

As for AppVeyor errors, we don't have the last results for v3.

**Comment by yegorich** [comment]
@bachp are there any actions needed from my side to get this PR merged?

Another question regarding CMake support. In Buildroot we setup following options in poco's [recipe](https://git.busybox.net/buildroot/tree/package/poco/poco.mk):

- --no-fpenvironment
- --no-wstring

They are not available as CMake options. I've looked at commit history and these options seems to be required any more. Do I understand this correctly?

**Comment by yegorich** [comment]
@bachp @aleks-f I've uploaded v4 some days ago. No changes just rebase on top of origin/develop. Appveyor build errors don't seem to have something with m changes.

**Comment by bachp** [comment]
@yegorich I don't think there are any equivalent options to the ones you mentioned. Also I think the probably just set some defines like `POCO_NO_FPENVIRONMENT`. If you read [here](https://github.com/pocoproject/poco/blob/d07dd731f360c24e61c30e05ddda60931ecef4e9/configure#L54) I guess they might not be needed anymore in most cases.

**Comment by zosrothko** [comment]
@yegorich the options

- --no-fpenvironment
- --no-wstring

are needed for building Poco with CMake on Cygwin platform

**Comment by bachp** [comment]
@zosrothko So it might be worth to add them. But this should be a separate PR.

File: Foundation/testsuite/CMakeLists.txt [hunk 1]
```diff
@@ -49,13 +49,11 @@ set_target_properties( TestApp PROPERTIES DEBUG_POSTFIX "") # don't use DEBUG_PO
 set_target_properties( TestApp PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY} )
 target_link_libraries( TestApp PocoFoundation)
 
-if(NOT POCO_STATIC)
+if(BUILD_SHARED_LIBS)
 # TestLibrary
-add_library( TestLibrary SHARED src/TestLibrary.cpp src/TestPlugin.cpp src/TestPlugin.h )
+add_library( TestLibrary src/TestLibrary.cpp src/TestPlugin.cpp src/TestPlugin.h )
```
[the remark this answers is not in the document]
Comment (bachp) [reply Foundation/testsuite/CMakeLists.txt]: Sounds reasonable.

**Comment by Bjoe** [comment]
I rebase to latest develop branch. 
I add also a warning https://github.com/yegorich/poco/blob/13177f942da3465c124df38e0ef37968cac1079b/CMakeLists.txt#L91 when POCO_STATIC is set because this can be happen when you switch between the branches. Also, TestLibrary should always compiled as shared. I will see if all tests are green and then I will merge this puil request.

**Pull request merged**